[osx] error: default initialization of an object of const type

### Does this issue affect the `google-cloud-cpp` project?

Yes

### What version of `google-cloud-cpp` are you using?

a0f27355af654605b321e764a7dd9464c54340a3

### What compiler and version are you using?

```
$ cc -v
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin17.7.0
Thread model: posix
```

### What operating system and version are you using?

```
$ system_profiler SPSoftwareDataType
Software:

    System Software Overview:

      System Version: macOS 10.13.6 (17G65)
      Kernel Version: Darwin 17.7.0
```

### What were you trying to do?

Compile

### What was the compilation error?

```
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++   -I/Users/ras0219/vcpkg/buildtrees/google-cloud-cpp/src/64c54340a3-41a12c0971 -I. -isystem /Users/ras0219/vcpkg/installed/x64-osx/include -fPIC   -isysroot /Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.11.sdk   -Wall -Werror -std=gnu++11 -MD -MT google/cloud/CMakeFiles/future_generic_test.dir/future_generic_test.cc.o -MF google/cloud/CMakeFiles/future_generic_test.dir/future_generic_test.cc.o.d -o google/cloud/CMakeFiles/future_generic_test.dir/future_generic_test.cc.o -c /Users/ras0219/vcpkg/buildtrees/google-cloud-cpp/src/64c54340a3-41a12c0971/google/cloud/future_generic_test.cc
/Users/ras0219/vcpkg/buildtrees/google-cloud-cpp/src/64c54340a3-41a12c0971/google/cloud/future_generic_test.cc:462:21: error: default initialization of an object of const type 'const future<int>' without a user-provided default constructor
  future<int> const f;
                    ^
/Users/ras0219/vcpkg/buildtrees/google-cloud-cpp/src/64c54340a3-41a12c0971/google/cloud/future_generic_test.cc:462:22: note: add an explicit initializer to initialize 'f'
  future<int> const f;
                     ^
                     {}
1 error generated.
```

--- a/.travis.yml
+++ b/.travis.yml
@@ -23,6 +23,7 @@ matrix:
     - # This is the only macOS entry in the matrix.  It is disabled on pull
       # requests because Travis often has long backlogs for macOS.
       os: osx
+      osx_image: xcode7.3
       compiler: clang
       env:
         OPENSSL_ROOT_DIR=/usr/local/opt/libressl

--- a/google/cloud/future_generic.h
+++ b/google/cloud/future_generic.h
@@ -37,8 +37,8 @@ class future final : private internal::future_base<T> {
   using shared_state_type =
       typename internal::future_base<T>::shared_state_type;
 
-  // workaround Apple Clang-7xx series bug, if we use `= default` here the
-  // compiler believes there is no default constructor defined :shrug:
+  // workaround Apple Clang-7xx series bug, if we use `= default` here, the
+  // compiler believes there is no default constructor defined. :shrug:
   future() noexcept {}
 
   /**

--- a/google/cloud/future_generic.h
+++ b/google/cloud/future_generic.h
@@ -37,7 +37,9 @@ class future final : private internal::future_base<T> {
   using shared_state_type =
       typename internal::future_base<T>::shared_state_type;
 
-  future() noexcept = default;
+  // workaround Apple Clang-7xx series bug, if we use `= default` here the
+  // compiler believes there is no default constructor defined :shrug:
+  future() noexcept {}
 
   /**
    * Creates a new future that unwraps @p rhs.

--- a/google/cloud/future_void.h
+++ b/google/cloud/future_void.h
@@ -37,7 +37,9 @@ class future<void> final : private internal::future_base<void> {
   using shared_state_type =
       typename internal::future_base<void>::shared_state_type;
 
-  future() noexcept = default;
+  // workaround Apple Clang-7xx series bug, if we use `= default` here the
+  // compiler believes there is no default constructor defined :shrug:
+  future() noexcept {}
 
   /**
    * Creates a new future that unwraps @p rhs.

--- a/google/cloud/future_void.h
+++ b/google/cloud/future_void.h
@@ -37,8 +37,8 @@ class future<void> final : private internal::future_base<void> {
   using shared_state_type =
       typename internal::future_base<void>::shared_state_type;
 
-  // workaround Apple Clang-7xx series bug, if we use `= default` here the
-  // compiler believes there is no default constructor defined :shrug:
+  // workaround Apple Clang-7xx series bug, if we use `= default` here, the
+  // compiler believes there is no default constructor defined. :shrug:
   future() noexcept {}
 
   /**

--- a/google/cloud/internal/future_impl_test.cc
+++ b/google/cloud/internal/future_impl_test.cc
@@ -173,12 +173,12 @@ TEST(ContinuationVoidTest, SetValueCallsContinuation) {
   SUCCEED();
 }
 
-thread_local int execute_counter;
-
 class TestContinuation : public continuation_base {
  public:
-  TestContinuation() = default;
-  void execute() override { ++execute_counter; }
+  TestContinuation(int* r) : execute_counter(r) {}
+  void execute() override { (*execute_counter)++; }
+
+  int* execute_counter;
 };
 
 TEST(FutureImplVoid, SetValue) {
@@ -244,9 +244,9 @@ TEST(FutureImplVoid, SetContinuation) {
   future_shared_state<void> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
-  execute_counter = 0;
+  int execute_counter = 0;
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
   EXPECT_EQ(0, execute_counter);
   EXPECT_FALSE(shared_state.is_ready());
   shared_state.set_value();
@@ -260,13 +260,15 @@ TEST(FutureImplVoid, SetContinuationAlreadySet) {
   future_shared_state<void> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
+  int execute_counter = 0;
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
 
   ExpectFutureError(
       [&] {
         shared_state.set_continuation(
-            google::cloud::internal::make_unique<TestContinuation>());
+            google::cloud::internal::make_unique<TestContinuation>(
+                &execute_counter));
       },
       std::future_errc::future_already_retrieved);
 }
@@ -275,11 +277,11 @@ TEST(FutureImplVoid, SetContinuationAlreadySatisfied) {
   future_shared_state<void> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
-  execute_counter = 0;
+  int execute_counter = 0;
   shared_state.set_value();
   EXPECT_EQ(0, execute_counter);
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
   EXPECT_EQ(1, execute_counter);
 
   shared_state.get();
@@ -407,9 +409,9 @@ TEST(FutureImplInt, SetContinuation) {
   future_shared_state<int> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
-  execute_counter = 0;
+  int execute_counter = 0;
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
   EXPECT_EQ(0, execute_counter);
   EXPECT_FALSE(shared_state.is_ready());
   shared_state.set_value(42);
@@ -423,12 +425,14 @@ TEST(FutureImplInt, SetContinuationAlreadySet) {
   future_shared_state<int> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
+  int execute_counter = 0;
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
   ExpectFutureError(
       [&] {
         shared_state.set_continuation(
-            google::cloud::internal::make_unique<TestContinuation>());
+            google::cloud::internal::make_unique<TestContinuation>(
+                &execute_counter));
       },
       std::future_errc::future_already_retrieved);
 }
@@ -437,11 +441,11 @@ TEST(FutureImplInt, SetContinuationAlreadySatisfied) {
   future_shared_state<int> shared_state;
   EXPECT_FALSE(shared_state.is_ready());
 
-  execute_counter = 0;
+  int execute_counter = 0;
   shared_state.set_value(42);
   EXPECT_EQ(0, execute_counter);
   shared_state.set_continuation(
-      google::cloud::internal::make_unique<TestContinuation>());
+      google::cloud::internal::make_unique<TestContinuation>(&execute_counter));
   EXPECT_EQ(1, execute_counter);
 
   EXPECT_EQ(42, shared_state.get());

